File: runtime/src/host/bundle_manager.rs
use std::collections::BTreeMap;

use async_trait::async_trait;

use crate::{common::crypto::hash::Hash, protocol::Protocol};

use super::{host_rpc_call, Error};

/// Name of the local RPC endpoint for the bundle manager.
pub const LOCAL_RPC_ENDPOINT_BUNDLE_MANAGER: &str = "bundle-manager";

/// Name of the BundleWrite method.
pub const METHOD_BUNDLE_WRITE: &str = "BundleWrite";
/// Name of the BundleAdd method.
pub const METHOD_BUNDLE_ADD: &str = "BundleAdd";
/// Name of the BundleRemove method.
pub const METHOD_BUNDLE_REMOVE: &str = "BundleRemove";
/// Name of the BundleList method.
pub const METHOD_BUNDLE_LIST: &str = "BundleList";

/// Name of the special label that identifies the instance.
pub const LABEL_INSTANCE_ID: &str = "net.oasis.instance_id";

/// Bundle manager interface.
#[async_trait]
pub trait BundleManager: Send + Sync {
    /// Request to host to write a chunk of the bundle to a temporary file.
    ///
    /// The `PermissionBundleAdd` permission is required to call this method.
    async fn bundle_write(&self, args: BundleWriteRequest) -> Result<BundleWriteResponse, Error>;

    /// Request to host to add a specific bundle to the host.
    ///
    /// The `PermissionBundleAdd` permission is required to call this method.
    async fn bundle_add(&self, args: BundleAddRequest) -> Result<BundleAddResponse, Error>;

    /// Request to host to remove a specific component. Only ROFL components added by this component
    /// can be removed.
    ///
    /// The `PermissionBundleRemove` permission is required to call this method.
    async fn bundle_remove(&self, args: BundleRemoveRequest)
        -> Result<BundleRemoveResponse, Error>;

    /// Request to host to list all bundles.
    ///
    /// The `PermissionBundleAdd` permission is required to call this method.
    async fn bundle_list(&self, args: BundleListRequest) -> Result<BundleListResponse, Error>;
}

#[async_trait]
impl BundleManager for Protocol {
    async fn bundle_write(&self, args: BundleWriteRequest) -> Result<BundleWriteResponse, Error> {
        host_rpc_call(
            self,
            LOCAL_RPC_ENDPOINT_BUNDLE_MANAGER,
            METHOD_BUNDLE_WRITE,
            args,
        )
        .await
    }

    async fn bundle_add(&self, args: BundleAddRequest) -> Result<BundleAddResponse, Error> {
        host_rpc_call(
            self,
            LOCAL_RPC_ENDPOINT_BUNDLE_MANAGER,
            METHOD_BUNDLE_ADD,
            args,
        )
        .await
    }

    async fn bundle_remove(
        &self,
        args: BundleRemoveRequest,
    ) -> Result<BundleRemoveResponse, Error> {
        host_rpc_call(
            self,
            LOCAL_RPC_ENDPOINT_BUNDLE_MANAGER,
            METHOD_BUNDLE_REMOVE,
            args,
        )
        .await
    }

    async fn bundle_list(&self, args: BundleListRequest) -> Result<BundleListResponse, Error> {
        host_rpc_call(
            self,
            LOCAL_RPC_ENDPOINT_BUNDLE_MANAGER,
            METHOD_BUNDLE_LIST,
            args,
        )
        .await
    }
}

/// Request to host to write a chunk of the bundle to a temporary file.
///
/// The `PermissionBundleAdd` permission is required to call this method.
#[derive(Clone, Debug, Default, cbor::Encode, cbor::Decode)]
pub struct BundleWriteRequest {
    /// Temporary file name to use on the host while writing the bundle.
    pub temporary_name: String,
    /// Optional flag which specifies that the temporary file should be recreated. If the file
    /// exists and this flag is set to true, it will be truncated. If the flag is set to false, any
    /// content will be appended to the existing file.
    pub create: bool,
    /// Data that should be appended to the temporary file.
    pub data: Vec<u8>,
}

/// Response form the BundleWrite method.
#[derive(Clone, Debug, Default, cbor::Encode, cbor::Decode)]
pub struct BundleWriteResponse {}

/// Request to host to add a specific bundle to the host.
///
/// The `PermissionBundleAdd` permission is required to call this method.
#[derive(Clone, Debug, Default, cbor::Encode, cbor::Decode)]
pub struct BundleAddRequest {
    /// Temporary file name to read the bundle from. The file must have previously been created by
    /// using `BundleWriteRequest`.
    ///
    /// The file must be a valid bundle.
    pub temporary_name: String,
    /// Expected hash of the manifest contained inside the bundle.
    pub manifest_hash: Hash,
    /// Labels to tag the bundle with.
    ///
    /// Note that the host will assign a random component identifier to these components, so one
    /// should use labels to later be able to find them.
    ///
    /// Use the special `LABEL_INSTANCE_ID` label to specify a deterministic instance ID.
    pub labels: BTreeMap<String, String>,
}

/// Response form the BundleAdd method.
#[derive(Clone, Debug, Default, cbor::Encode, cbor::Decode)]
pub struct BundleAddResponse {}

/// Request to host to remove a specific component. Only ROFL components added by this component can
/// be removed.
///
/// The `PermissionBundleRemove` permission is required to call this method.
#[derive(Clone, Debug, Default, cbor::Encode, cbor::Decode)]
pub struct BundleRemoveRequest {
    /// Labels to filter the components by.
    pub labels: BTreeMap<String, String>,
}

/// Response form the BundleRemove method.
#[derive(Clone, Debug, Default, cbor::Encode, cbor::Decode)]
pub struct BundleRemoveResponse {}

/// Request to host to list all bundles.
///
/// The `PermissionBundleAdd` permission is required to call this method.
#[derive(Clone, Debug, Default, cbor::Encode, cbor::Decode)]
pub struct BundleListRequest {
    /// Labels to filter the components by.
    pub labels: BTreeMap<String, String>,
}

/// Response from host to list all bundles.
#[derive(Clone, Debug, Default, cbor::Encode, cbor::Decode)]
pub struct BundleListResponse {
    /// The resulting bundles.
    #[cbor(optional)]
    pub bundles: Vec<BundleInfo>,
}

/// Bundle information.
#[derive(Clone, Debug, Default, cbor::Encode, cbor::Decode)]
pub struct BundleInfo {
    /// Hash of the manifest.
    pub manifest_hash: Hash,
    /// List of all components in this bundle.
    pub components: Vec<ComponentInfo>,
    /// Labels assigned to this bundle.
    pub labels: BTreeMap<String, String>,
}

/// Component information.
#[derive(Clone, Debug, Default, cbor::Encode, cbor::Decode)]
pub struct ComponentInfo {
    /// Component name.
    pub name: String,
}

File: go/runtime/rofl/api/bundle_manager.go
package api

import "github.com/oasisprotocol/oasis-core/go/common/crypto/hash"

const (
	// LocalRPCEndpointBundleManager is the name of the local RPC endpoint for the bundle manager.
	LocalRPCEndpointBundleManager = "bundle-manager"

	// MethodBundleWrite is the name of the BundleWrite method.
	MethodBundleWrite = "BundleWrite"
	// MethodBundleAdd is the name of the BundleAdd method.
	MethodBundleAdd = "BundleAdd"
	// MethodBundleRemove is the name of the BundleRemove method.
	MethodBundleRemove = "BundleRemove"
	// MethodBundleList is the name of the BundleList method.
	MethodBundleList = "BundleList"
)

// BundleWriteRequest is a request to host to store a chunk of the bundle.
//
// The `PermissionBundleAdd` permission is required to call this method.
type BundleWriteRequest struct {
	// TemporaryName is a temporary name to identify the chunk later.
	TemporaryName string `json:"temporary_name"`
	// Create is the optional flag which specifies that the bundle should be recreated. If the
	// bundle exists and this flag is set to true, it will be truncated. If the flag is set to
	// false, any content will be appended to the existing bundle.
	Create bool `json:"create,omitempty"`
	// Data that should be appended to the bundle.
	Data []byte `json:"data"`
}

// BundleWriteResponse is the response from the BundleWrite method.
type BundleWriteResponse struct{}

// BundleAddRequest is a request to host to add a specific bundle to the host.
//
// The `PermissionBundleAdd` permission is required to call this method.
type BundleAddRequest struct {
	// TemporaryName is the temporary name to use to access the bundle. The chunks must have
	// previously been created by using `BundleWriteRequest`.
	TemporaryName string `json:"temporary_name"`
	// ManifestHash is the expected hash of the manifest contained inside the bundle.
	ManifestHash hash.Hash `json:"manifest_hash"`
	// Labels are the labels to tag the bundle with so it can later be found.
	Labels map[string]string `json:"labels"`
	// Volumes are the volumes to attach to the bundle.
	Volumes map[string]string `json:"volumes"`
}

// BundleAddResponse is the response from the BundleAdd method.
type BundleAddResponse struct{}

// BundleRemoveRequest is a request to host to remove specific bundles.
//
// The `PermissionBundleRemove` permission is required to call this method.
type BundleRemoveRequest struct {
	// Labels are the labels to filter the bundles by. All labels must match.
	Labels map[string]string `json:"labels"`
}

// BundleRemoveResponse is the response from the BundleRemove method.
type BundleRemoveResponse struct{}

// BundleListRequest is a request to host to list all bundles.
//
// The `PermissionBundleAdd` permission is required to call this method.
type BundleListRequest struct {
	// Labels are the labels to filter the bundles by. All labels must match.
	Labels map[string]string `json:"labels"`
}

// BundleListResponse is a response from host to list all bundles.
type BundleListResponse struct {
	// Bundles are the resulting bundles.
	Bundles []*BundleInfo `json:"bundles,omitempty"`
}

// BundleInfo is the bundle information.
type BundleInfo struct {
	// ManifestHash is the hash of the manifest.
	ManifestHash hash.Hash `json:"manifest_hash"`
	// Components is a list of all components in this bundle.
	Components []*ComponentInfo `json:"components"`
	// Labels is a set of labels assigned to this bundle.
	Labels map[string]string `json:"labels,omitempty"`
}

// ComponentInfo is the component information.
type ComponentInfo struct {
	// Name is the component name.
	Name string `json:"name"`
}
